### ISSUE TITLE
torch.compile generates wrong code on CPU and compiled code replaces original function

### ISSUE THREAD

**Issue by gau-nernst** [pytorch/pytorch#126848]
### 🐛 Describe the bug

```python
import torch
import torch._inductor.config

torch._inductor.config.force_disable_caches = True  # doesn't seem to help

print(torch.__version__)

def func(tensor: torch.Tensor) -> torch.Tensor:
    return tensor * 2.0 ** (-127 + 3)

x = torch.tensor(0.1875)
print("Eager (before):", func(x))
print("Compile:", torch.compile(func)(x))
print("Eager (after):", func(x))
```

Output

```
2.4.0.dev20240521
Eager (before): tensor(8.8162e-39)
Compile: tensor(0.)
Eager (after): tensor(0.)
```

There are 2 separate issues
1. Compiled version produces wrong results (compare `Eager (before)` and `Compile` lines)
    - I compared the generated code (using `TORCH_LOGS="output_code"`). The generated C++ kernel is identical, only the accompanied header file is different.
2. After running `torch.compile()`, the original eager function produces wrong results like the compiled version (compare `Eager (before)` and `Eager (after)`)
    - Seems like some caching issue that shouldn't happen?
 
I only observe this issue for nightly build. v2.3 is not affected.

This bug was discovered while working on https://github.com/pytorch/ao/pull/248

@msaroufim

### Versions

```
Collecting environment information...
PyTorch version: 2.4.0.dev20240521
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.10.14 (main, May  6 2024, 19:42:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-1058-aws-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      46 bits physical, 48 bits virtual
CPU(s):                             4
On-line CPU(s) list:                0-3
Thread(s) per core:                 2
Core(s) per socket:                 2
Socket(s):                          1
NUMA node(s):                       1
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              85
Model name:                         Intel(R) Xeon(R) Platinum 8175M CPU @ 2.50GHz
Stepping:                           4
CPU MHz:                            3103.782
BogoMIPS:                           4999.99
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          64 KiB
L1i cache:                          64 KiB
L2 cache:                           2 MiB
L3 cache:                           33 MiB
NUMA node0 CPU(s):                  0-3
Vulnerability Gather data sampling: Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:        KVM: Mitigation: VMX unsupported
Vulnerability L1tf:                 Mitigation; PTE Inversion
Vulnerability Mds:                  Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:             Vulnerable
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves ida arat pku ospke

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.2
[pip3] pytorch-lightning==2.2.3
[pip3] torch==2.2.1+cu121
[pip3] torch-xla==2.3.0
[pip3] torchao==0.2.0
[pip3] torchaudio==2.2.0.dev20240521
[pip3] torchmetrics==1.3.1
[pip3] torchvision==0.17.1+cu121
[pip3] triton==2.2.0
[conda] blas                      1.0                         mkl  
[conda] libjpeg-turbo             2.0.0                h9bf148f_0    pytorch-nightly
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-service               2.4.0           py310h5eee18b_1  
[conda] mkl_fft                   1.3.8           py310h5eee18b_0  
[conda] mkl_random                1.2.4           py310hdb19cb5_0  
[conda] numpy                     1.26.2                   pypi_0    pypi
[conda] numpy-base                1.26.4          py310hb5e798b_0  
[conda] pytorch                   2.4.0.dev20240521 py3.10_cuda12.1_cudnn8.9.2_0    pytorch-nightly
[conda] pytorch-cuda              12.1                 ha16c6d3_6    pytorch-nightly
[conda] pytorch-lightning         2.2.3                    pypi_0    pypi
[conda] pytorch-mutex             1.0                        cuda    pytorch-nightly
[conda] torch                     2.2.1+cu121              pypi_0    pypi
[conda] torch-xla                 2.3.0                    pypi_0    pypi
[conda] torchao                   0.2.0                     dev_0    <develop>
[conda] torchaudio                2.2.0.dev20240521     py310_cu121    pytorch-nightly
[conda] torchmetrics              1.3.1                    pypi_0    pypi
[conda] torchtriton               3.0.0+45fff310c8           py310    pytorch-nightly
[conda] torchvision               0.17.1+cu121             pypi_0    pypi
[conda] triton                    2.2.0                    pypi_0    pypi
```

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @bdhirsh @anijain2305 @chauhang

**Comment by msaroufim** [comment]
High pri because of silent correctness issue

**Comment by jgong5** [comment]
It might be caused by the change of the floating point state during inductor compilation. The result of the test exceeds the normal range of the floating point but in the range of the denormals. If we do flush denormals to zero, the returned value of 0 is expected.

**Comment by gau-nernst** [comment]
Confirm that flushing denormal to zero is the culprit. However, the compiled version always flushes denormal to zero. And whenever I call the compiled version, flushing denormal is set to True again.

```python
import torch

print(torch.__version__)

def func(tensor: torch.Tensor) -> torch.Tensor:
    return tensor * 2.0 ** (-127 + 3)

x = torch.tensor(0.1875)
print("Eager (before):", func(x))

func_compiled = torch.compile(func)
torch.set_flush_denormal(False)
print("Compile:", func_compiled(x))

print("Eager (after 1):", func(x))
torch.set_flush_denormal(False)
print("Eager (after 2):", func(x))
```

```
2.4.0.dev20240521
Eager (before): tensor(8.8162e-39)
Compile: tensor(0.)
Eager (after 1): tensor(0.)
Eager (after 2): tensor(8.8162e-39)
```

For the PR I'm working on, I'm using FP6, which doesn't have a lot of resolution. Flushing denormal is definitely going to be a problem.

**Comment by leslie-fang-intel** [comment]
Inductor CPP backend turns on the GCC flag of `-ffast-math`.

According to this link: https://stackoverflow.com/questions/47101001/can-i-temporarily-enable-ftz-and-daz-floating-point-modes-for-a-thread

> Note that linking with -ffast-math in gcc will include a CRT function that sets DAZ/FTZ before main().

**Comment by jgong5** [comment]
> Inductor CPP backend turns on the GCC flag of `-ffast-math`.
> 
> According to this link: https://stackoverflow.com/questions/47101001/can-i-temporarily-enable-ftz-and-daz-floating-point-modes-for-a-thread

Perhaps we need to save/restore these fp flags during inductor kernel calls...

> For the PR I'm working on, I'm using FP6, which doesn't have a lot of resolution. Flushing denormal is definitely going to be a problem.

@gau-nernst Can you elaborate more on this? Do you mean the normal fp32 data range cannot cover that of fp6? Please note that the computation with denormals is slooow..

**Comment by gau-nernst** [comment]
I'm working on FP32<->FP6 dtype conversion. I'm implementing this by correcting exponent bias via FP32 multiplication, so it will handle denormal values correctly for me. Then I only need to bit-shift the sign, exponent, and mantissa bits to their correct positions. There is no need to consider 2 separate cases (1) FP32 normal -> FP6 normal, and (2) FP32 normal -> FP6 denormal (FP32 denormal always underflows to 0 in FP6). The short version of the code is below

```python
# no rounding logic for brevity
def fp32_to_fp6(tensor: Tensor) -> Tensor:
    # correct exponent bias. this also handles subnormal numbers correctly
    tensor = tensor * 2.0 ** (-127 + 3)
    bits = tensor.view(torch.int32)

    sign = ((bits >> 31) & 0x1) << 5
    exp_and_man = (bits >> 21) & 0x1F
    result = sign | exp_and_man
    return result.to(torch.uint8)  # stored as uint8 data

def fp6_to_fp32(tensor: Tensor) -> Tensor:
    bits = tensor.to(torch.int32)  # bit extension
    sign = bits >> 5 << 31
    exp_and_man = (bits & 0x1F) << 21
    results = sign | exp_and_man

    results = results.view(torch.float32)
    return results * 2.0 ** (127 - 3)  # exponent bias correction
```

Therefore, it's important for my feature that denormal numbers are handled correctly.

I observe that this problem only occurs with PyTorch nightly, not with 2.3 and 2.2.2. Perhaps some compile flags were changed since then? A quick GitHub search shows that `-ffast-math` is already enabled since 2.2.2 (not sure if I'm looking at the right place), but this problem didn't happen with 2.2.2 and 2.3?

https://github.com/pytorch/pytorch/blob/v2.3.0/torch/_inductor/codecache.py#L1245

https://github.com/pytorch/pytorch/blob/v2.2.2/torch/_inductor/codecache.py#L1195

But anyway, I think torch.compile should not mess with DAZ/FTZ flags.

**Comment by jgong5** [comment]
> I'm working on FP32<->FP6 dtype conversion. I'm implementing this by correcting exponent bias via FP32 multiplication, so it will handle denormal values correctly for me. Then I only need to bit-shift the sign, exponent, and mantissa bits to their correct positions. There is no need to consider 2 separate cases (1) FP32 normal -> FP6 normal, and (2) FP32 normal -> FP6 denormal (FP32 denormal always underflows to 0 in FP6). The short version of the code is below

Just curious, when you implement fp32 -> fp6, why not just doing bits operations without arithmetic ops? That can avoid the need of manipulating the denormals which are inefficient.

> But anyway, I think torch.compile should not mess with DAZ/FTZ flags.

Yes. This is what we should fix. WIP.

**Comment by gau-nernst** [comment]
I did have an implementation with bit manipulation for CPU in C++. But this cannot be implemented efficiently in PyTorch (then speed up with torch.compile) since it requires if-else branches - 1 for FP6 normal numbers and 1 for FP6 denormal numbers. To implement it in PyTorch, I have to calculate everything in both branches, then select the results with `torch.where()`. Using together with torch.compile, the C++ generated code still computes everything in both branches, instead of calculating what is needed in each branch. This slows down significantly.

One reason of not keeping this C++ implementation in torchao is to keep the code simple. On CUDA, the PyTorch+torch.compile fp32 multiplication implementation is fast (it's memory-bound) while still very simple. Re-using the same code for CPU would be ideal.

**Comment by leslie-fang-intel** [comment]
Objdump the generated code with PyTorch: 5961e23e76a83640170916b5cecd5090512c0b74 (similar behavior as PyTorch 2.3
), we also see these lines to set the mxcsr registers which add by the `-ffast_math` flag.
```
0000000000003c60 <set_fast_math>:
    3c60:	f3 0f 1e fa          	endbr64 
    3c64:	0f ae 5c 24 fc       	stmxcsr -0x4(%rsp)
    3c69:	81 4c 24 fc 40 80 00 	orl    $0x8040,-0x4(%rsp)
    3c70:	00 
    3c71:	0f ae 54 24 fc       	ldmxcsr -0x4(%rsp)
    3c76:	c3                   	ret    
    3c77:	66 0f 1f 84 00 00 00 	nopw   0x0(%rax,%rax,1)
    3c7e:	00 00 
```

The interesting thing is mxcsr register has been reset to  0x1fab before actually execute the kernel.

![image](https://github.com/pytorch/pytorch/assets/53841472/7d6df7e3-ac6e-4c44-8f8b-b12f471b4552)

The behavior changed start with 285c93d64db23b5cf495c52458d9b3120b96ee05.

**Comment by Valentine233** [comment]
Instead of using `-ffast-math`, the issue can be avoided by using its detailed options: `-fno-math-errno, -funsafe-math-optimizations, -ffinite-math-only, -fno-rounding-math, -fno-signaling-nans, -fcx-limited-range and -fexcess-precision=fast`. We need to validate for the change.

In addition, it is also suggested by Meta to entirely remove `-ffast-math` because of the potential accuracy issues. Also need the validation.

**Comment by msaroufim** [comment]
@Valentine233 any update on this? Can we not use fast math lol

**Comment by Valentine233** [comment]
> @Valentine233 any update on this? Can we not use fast math lol

Sorry, we haven't started the validation yet, because of the machine changes. cc @leslie-fang-intel

**Comment by Valentine233** [comment]
Depends on https://github.com/pytorch/pytorch/issues/124245.

**Comment by pianpwk** [comment]
(issue scrubbing) @gau-nernst this seems to work on nightly? Could you confirm if it's still an issue?

**Comment by masnesral** [comment]
> (issue scrubbing) @gau-nernst this seems to work on nightly? Could you confirm if it's still an issue?

Gonna assume this is fixed, but feel free to reopen if we're mistaken.
